Loop/Managers/LoopDataManager.swift

```diff
@@ -226,7 +226,12 @@ final class LoopDataManager: LoopSettingsAlerterDelegate {
             self.carbsOnBoard = nil
             self.insulinEffect = nil
         }
-
+        
+        if newValue.maximumBasalRatePerHour != oldValue.maximumBasalRatePerHour {
```

Comment:
Okay, now I'm confused. 

If the user reduced their max basal (per above), then one of two things can happen.
1) The max basal is *still above* (or equal to) an active temp basal. In this case, the active temp basal wouldn't change anyhow so no need to loop again.
2) The max max is *below* an active temp basal. This case is already handled by `LoopDataManager.cancelActiveTempBasal` where the active temp basal is cancelled.

What am I missing?

Now, if there was a concern where the active temp basal was *equal to* the OLD max basal (and, so, theoretically could go even higher with a new loop), then that pretty much goes for any other settings you change. Changing target range, basal rates, carb ratios, insulin sensitivities will *all* potentially change the current active temp basal (or lack thereof).

Reply:
I think @ps2 has to chime in here, but my understanding is just because you cancel the current temp basal, it doesn't mean that the proper basal rate is recalculated unless `loop()` is called.

Reply:
Oh, then if it is **only** when the an active temp basal is cancelled (after the max basal is set *lower* than the active temp basal), then it should key off that explicitly, not just *any* change to the max basal. In that case, I think it might make sense to invoke `loop()` from within the completion handler of `enactRecommendedAutomaticDose` within `cancelActiveTempBasal`.

Reply:
Hm... Don't we also want to recalculate even if we don't cancel it?

Reply:
If we don't cancel it, then it falls back into the same "we don't force a new loop on any other settings changes" category. I get it for a cancelTempBasal that is due to a lowered max basal setting change since you just canceled the active temp basal.

Reply:
Well, we only would want to do that if `cancelActiveTempBasal` succeeds.  Also, I will note that `cancelActiveTempBasal` is called in other places, so I would probably recommend instead putting it in `maxTempBasalSavePreflight`, probably _after_ the `completion` block is called.  Still, I am not a huge fan of having `maxTempBasalSavePreflight` have `loop()` as a side-effect, but again I think this is @ps2 's call.

Reply:
Well, for unrelated reasons (namely, complete and accurate dosing decision records), I'm adding a "reason" to the `cancelActiveTempBasal` function. One of those reasons is `.maximumBasalRateChanged` so we could key off that. :)

Reply:
@darinkrauss , it's a good point that we don't currently call `loop` on other settings changes, though maybe we should. And I hear @rickpasetto that that is beyond the scope of what we're trying to tackle with this. 

I'm not sure Darin's idea about keying off `.maximumBasalRateChanged`,  works, as we have this set of steps as a goal:

1. User changes max temp basal below currently running temp basal
2. tempBasal is canceled (async)
3. updated settings are persisted iff temp basal cancellation succeeds.
4. loop() is triggered

Steps 2,3, and 4 should all be done within LDM as a single async operation, imo. But I think right now the UI triggers 2 and *then* tries to save. At the point of save, we don't know where it came from, so adding a `loop` call there would be somewhat indiscriminate, which is what I understand Darin's concern to be.

Can we change the protocols so that the UI doesn't do a "preflight" and then save? But rather a single "save" that ultimately triggers all this to happen in LDM as a single call, with the appropriate logic?

Reply:
@ps2 the `save` is not only indiscriminate, it also "can't" fail.  So, it is possible, I think, to do the `loop` in the preflight, after the `completion` is called.  Look at the commit I added, and LMK if you think this is suitable, without going through changing the protocols.

Reply:
But aren't you causing a loop() with the old values, then?  Seems like we should do this all in LDM, as one call.

Reply:
Ah, I see what you're saying now, @ps2 .   

Ok, well, to do the refactor you are asking for will be somewhat complicated, unfortunately, because now (in 1.2.0) the preflight process also involves calling `syncDeliveryLimits` (in DDM), and only after _that_ succeeds do we unwind all the way back up to `DeliveryLimitsEditor` to proceed with `save`.  This all still has to interact with the `SavingMechanism` in `DeliveryLimitsEditor` and user authentication.  In other words, this is all getting pretty tangled.  While I agree it is a bit of a mess, cleaning it up seems like a big lift, compared to what this PR originally was, which was a few lines of code that failed to merge to 1.2.0.  

I'd also like to humbly note that we are considering [yet] another refactoring change to code that was originally just meant to implement [LOOP-3856](https://tidepool.atlassian.net/browse/LOOP-3856), which, while important as a risk mitigation, was a bit of a corner case.

Just want to step back a bit: The implementation originally assumed that calling `loop()`, even an "extra" time, was not harmful at all.  Is that not true?

Reply:
Calling loop an extra time shouldn't do anything worse than just being a waste of battery due to extra comms. The issue I was pointing out was that it wouldn't be running with the updated values, so not really worth doing at that point.

The biggest part of this mitigation is getting the current high temp basal canceled. I think triggering a loop is a nicety, but not critical by any means. So if we just drop the requirement to run loop, I think we'll still be addressing the main risk.

Reply:
@ps2 Yes, I was agreeing with the issue you pointed out (I've reverted my proposed commit).  I'm a bit confused, now, though...and this may just be a misunderstanding on my part.  I thought that the `loop` was more than a nicety, but that waiting for the next loop cycle was a risk we didn't want.   Am I mistaken?  If that's the case, then is https://tidepool.atlassian.net/browse/LOOP-3915 even a bug then?

Reply:
I think we could argue that it's not needed. I don't see a risk associated with this ticket; I think canceling the temp basal fixes the main risk that existed. Let's push back on this one and see if it really needs to be addressed.

Reply:
Ok I'll bring it up at standup.  Meanwhile, separately, I'll take a look to see if doing some rework could clean up "save".